orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -869,89 +869,217 @@ common::Status HandleSharedInitializer(Graph& graph,
+                                 int nstages,
+                                 std::vector<std::pair<int, int>>& messages,
+                                 std::vector<Node*>& send_nodes,
+                                 std::vector<Node*>& recv_nodes) {
+
+  // forward_messages[s]: all the tensors sent by stage s while executing
+  // forward computation.
+  std::vector<std::set<const NodeArg*>> forward_messages(nstages);
+  // TODO(jufranc): once we start using this function on the training graph,
+  // we need to keep backward_messages[s] too.
+
+  // Tensors that need to be sent from one device to the other.
+  // TODO(jufranc): Should we consider weights here too? 
+  // forwarded_tensors[i] = {t, {stage of producer, stage of the last consumer}}
+  std::vector<std::pair<const NodeArg*, std::pair<int, int>>> forwarded_tensors;
+
+  // All the tensors that are produced and consumed in the graph.
+  auto initial_node_args = GetAllNodeArgs(graph);
+
+  // We create all the tensor replicas in advance using this fuction.
+  // A tensor produced in stage r and consumed in stage r', such that r' > r,
+  // will have a replica in all stages r'', such that r'' > r and r'' < r'.
+  // tensor_replicas[t][r] contains a pointer to the the replica of t in stage r, 
+  // it if exists, or to itself if r is the stage of the producer of r.
+  std::map<const NodeArg*, std::vector<NodeArg*>> tensor_replicas;
+  auto create_tensor_replica = [&tensor_replicas, &graph](const NodeArg* tensor,
+                                                          int consumer_stage) {
+    NodeArg& new_receive_output = CreateNodeArg(graph, *tensor);
+    const auto* old_shape = tensor->Shape();
+    if (old_shape != nullptr) {
+      new_receive_output.SetShape(*old_shape);
+    }
+    // Add value info for this newly added receive_output, for shape propagation
+    // when training this partition.
+    graph.AddValueInfo(&new_receive_output);
+    tensor_replicas[tensor][consumer_stage] = &new_receive_output;
+  };
+
+  // Checks whether the tensor is produced and consumed in the forward stage of
+  // the computation.
+  auto is_forward = [](const int producer_stage, const int consumer_stage) {
+    return producer_stage < consumer_stage;
+  };
+
+  // Checks whether the tensor is produced and consumed in the backward stage of
+  // the computation. 
+  auto is_backward = [](const int producer_stage, const int consumer_stage) {
+    return producer_stage > consumer_stage;
+  };
+
+  // Find tensors that need to be sent and forwarded.
+  for (const NodeArg* node_arg : initial_node_args) {
+    // Initialize tensor_replicas data structure.
+    auto inserted = tensor_replicas.emplace(
+                      std::make_pair(node_arg, std::vector<NodeArg*>(nstages)));
+    auto& replicas = (*inserted.first).second;
+
+    // TODO: for now we pretend that inputs are produced in stage 0,
+    // but I need to double check how they are handled.
```

Comment:
I guess you mean `if there is no producer, then that tensor must be a graph input and all graph inputs are created at pipeline stage 0.`? If yes, shall we remove
```
assert(producer_node != nullptr);
```
because graph inputs don't have producer.

Reply:
Yes, this is what I meant. However, GetAllNodeArgs should not include inputs of the graph -- a tensor without producer here should be an error. So, what I need is what you suggested in the comment below. 

```
Node* producer_node = graph.GetMutableProducerNode(node_arg->Name());
assert(producer_node != nullptr);
int producer_stage = op_to_stage.find(producer_node)->second;
```